internal/metricdata/subscriber.go

```diff
@@ -0,0 +1,215 @@
+	return smContextActive
+}
+
+func GetSMContextCount() uint64 {
+	atomic.AddUint64(&smContextCount, 1)
+	return smContextCount
+}
+
+func HandleSubscriberEvent(subsData *metricinfo.CoreSubscriberData, sourceNf metricinfo.NfType) {
+	switch subsData.Operation {
+	case metricinfo.SubsOpAdd:
+		storeSubscriber(&subsData.Subscriber, sourceNf)
+	case metricinfo.SubsOpMod:
+		updateSubscriber(&subsData.Subscriber, sourceNf)
+	case metricinfo.SubsOpDel:
+		deleteSubscriber(&subsData.Subscriber, sourceNf)
+	default:
+		logger.CacheLog.Errorf("unknown smf subsriber operation [%v] ", subsData.Operation)
+	}
+}
+
+func storeSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) error {
+	metricData.SubLock.Lock()
+	//defer metricData.SubLock.Unlock()
+	_, ok := metricData.Subscribers[sub.Imsi]
+	if !ok {
+		metricData.Subscribers[sub.Imsi] = sub
+
+		promclient.SetSmfSessStats(sub.SmfIp, sub.Slice, sub.Dnn, sub.UpfName, incSMContextActive())
+		logger.CacheLog.Debugf("storing subscriber with imsi [%s] ", sub.Imsi)
+		pushPrometheusCoreSubData(sub)
+		metricData.SubLock.Unlock()
+	} else {
+		metricData.SubLock.Unlock()
+		updateSubscriber(sub, sourceNf)
+	}
+
+	return nil
+}
+
+func updateSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) {
+	metricData.SubLock.Lock()
+	defer metricData.SubLock.Unlock()
+	if s, ok := metricData.Subscribers[sub.Imsi]; ok {
+
+		deletePrometheusCoreSubData(s)
+
+		if sourceNf == metricinfo.NfTypeSmf {
+			//SMF specific fields
+			fillSmfSubsriberData(sub, s)
+		} else if sourceNf == metricinfo.NfTypeAmf {
+			//AMF specific fields
+			fillAmfSubsriberData(sub, s)
+		}
+		pushPrometheusCoreSubData(s)
+	}
+}
+
+func deleteSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) error {
+	metricData.SubLock.Lock()
+	defer metricData.SubLock.Unlock()
+	imsi := sub.Imsi
+	s, ok := metricData.Subscribers[imsi]
+	if ok {
+		deletePrometheusCoreSubData(s)
+		s.SmfSubState = sub.SmfSubState
+		s.AmfSubState = sub.AmfSubState
+
+		//register disconnect state
+		pushPrometheusCoreSubData(s)
+		delete(metricData.Subscribers, imsi)
+
+		//register subscriber delete
+		deletePrometheusCoreSubData(s)
+		promclient.SetSmfSessStats(s.SmfIp, sub.Slice, sub.Dnn, sub.UpfName, decSMContextActive())
+		logger.CacheLog.Debugf("deleting subscriber with imsi [%s] ", imsi)
+	}
+
+	return nil
+}
+
+func GetSubscriber(key string) (*metricinfo.CoreSubscriber, error) {
+	metricData.SubLock.RLock()
+	defer metricData.SubLock.RUnlock()
+	if sub, ok := metricData.Subscribers[key]; ok {
+		return sub, nil
+	}
+	return nil, fmt.Errorf("subscriber with key [%v] not found ", key)
+}
+
+func GetSubscriberAll() []string {
+	imsis := []string{}
+	metricData.SubLock.RLock()
+	defer metricData.SubLock.RUnlock()
+
+	for imsi := range metricData.Subscribers {
+		imsis = append(imsis, imsi)
+	}
+
+	return imsis
+}
+
+func pushPrometheusCoreSubData(sub *metricinfo.CoreSubscriber) {
+	//Push to prometheus
```

Comment:
mis leading function name and comment. We are not using data to prometheus. We are updating local prometheus client registry. Prometheus will scrape/pull data at periodic interval.

Reply:
I meant pushing data to prometheus module. logical push and not actual push to Prometheus server. I shall correct the naming.